src/client/datascience/ipywidgets/notebookIPyWidgetCoordinator.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { inject, injectable } from 'inversify';
+import { CancellationToken, Disposable } from 'vscode';
+import { NotebookCommunication, NotebookDocument, NotebookKernel } from '../../../../types/vscode-proposed';
+import { IVSCodeNotebook } from '../../common/application/types';
+import { Cancellation } from '../../common/cancellation';
+import { createDeferred } from '../../common/utils/async';
+import { IServiceContainer } from '../../ioc/types';
+import { InteractiveWindowMessages, IPyWidgetMessages } from '../interactive-common/interactiveWindowTypes';
+import { INotebookKernelResolver } from '../notebook/types';
+import { CommonMessageCoordinator } from './commonMessageCoordinator';
+
+/**
+ * This class wires up VSC notebooks to ipywidget communications.
+ */
+@injectable()
+export class NotebookIPyWidgetCoordinator implements INotebookKernelResolver {
+    private messageCoordinators = new Map<string, Promise<CommonMessageCoordinator>>();
+    private attachedWebViews = new Map<string, { webviews: Set<string>; disposables: Disposable[] }>();
+    private disposables: Disposable[] = [];
+    constructor(
+        @inject(IServiceContainer) private readonly serviceContainer: IServiceContainer,
+        @inject(IVSCodeNotebook) private readonly notebookProvider: IVSCodeNotebook
+    ) {
+        notebookProvider.onDidCloseNotebookDocument(this.onDidCloseNotebook.bind(this));
+    }
+    public dispose(): void | undefined {
+        this.messageCoordinators.forEach((v) => v.then((c) => c.dispose()));
+        this.messageCoordinators.clear();
+    }
+    public resolveKernel(
+        _kernel: NotebookKernel,
+        document: NotebookDocument,
+        webview: NotebookCommunication,
+        token: CancellationToken
+    ): Promise<void> {
+        // Create a handler for this notebook if we don't already have one. Since there's one of the notebookMessageCoordinator's for the
+        // entire VS code session, we have a map of notebook document to message coordinator
+        let promise = this.messageCoordinators.get(document.uri.toString());
+        if (!promise) {
+            promise = CommonMessageCoordinator.create(document.uri, this.serviceContainer);
+            this.messageCoordinators.set(document.uri.toString(), promise);
+        }
+        return Cancellation.race(() => promise!.then(this.attachCoordinator.bind(this, document, webview)), token);
+    }
+
+    private onDidCloseNotebook(e: NotebookDocument) {
+        // See if this is the last copy of this document
+        if (!this.notebookProvider.notebookDocuments.find((d) => d.uri !== e.uri)) {
```

Comment:
Am I reading this wrong? 

Looks like if you can't find a document with a different URI in the list of open documents. Don't you want to not find a document with the same uri?

Reply:
Yeah this is wrong. It worked because I only ever had a single document open when testing.

Reply:
This was my only comment that was really an issue. Rest was all minor cleanup / suggestions.